apiserver/facades/client/client/statuslegacy_test.go

```diff
@@ -0,0 +1,1057 @@
+	release()
+
+	host := factory.MakeMachine(c, &testfactory.MachineParams{InstanceId: "1"})
+	factory.MakeMachineNested(c, host.Id(), nil)
+	lxdHost := factory.MakeMachineNested(c, host.Id(), nil)
+	factory.MakeMachineNested(c, lxdHost.Id(), nil)
+
+	conn := s.OpenControllerModelAPI(c)
+	client := apiclient.NewClient(conn, loggertesting.WrapCheckLog(c))
+	status, err := client.Status(context.Background(), nil)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(status.Machines, gc.HasLen, 1)
+	mStatus, ok := status.Machines[host.Id()]
+	c.Check(ok, jc.IsTrue)
+	c.Check(mStatus.Containers, gc.HasLen, 2)
+
+	mStatus, ok = mStatus.Containers[lxdHost.Id()]
+	c.Check(ok, jc.IsTrue)
+
+	c.Check(mStatus.Containers, gc.HasLen, 1)
+}
+
+func (s *statusUnitTestSuite) TestApplicationWithExposedEndpoints(c *gc.C) {
+	factory, release := s.NewFactory(c, s.ControllerModelUUID())
+	release()
+
+	charm := factory.MakeCharm(c, &testfactory.CharmParams{Name: "wordpress", URL: "ch:amd64/wordpress"})
+	app := factory.MakeApplication(c, &testfactory.ApplicationParams{Charm: charm})
+	err := app.MergeExposeSettings(map[string]state.ExposedEndpoint{
+		"": {
+			ExposeToSpaceIDs: []string{network.AlphaSpaceId},
+			ExposeToCIDRs:    []string{"10.0.0.0/24", "192.168.0.0/24"},
+		},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	conn := s.OpenControllerModelAPI(c)
+	client := apiclient.NewClient(conn, loggertesting.WrapCheckLog(c))
+	status, err := client.Status(context.Background(), nil)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(status, gc.NotNil)
+	appStatus, ok := status.Applications[app.Name()]
+	c.Assert(ok, gc.Equals, true)
+
+	c.Assert(appStatus.ExposedEndpoints, gc.DeepEquals, map[string]params.ExposedEndpoint{
+		"": {
+			ExposeToSpaces: []string{network.AlphaSpaceName},
+			ExposeToCIDRs:  []string{"10.0.0.0/24", "192.168.0.0/24"},
+		},
+	})
+}
+
+func defaultCharmOrigin(curlStr string) *state.CharmOrigin {
+	// Use ParseURL here in test until either the charm and/or application
+	// can easily provide the same data.
+	curl, _ := charm.ParseURL(curlStr)
+	var source string
+	var channel *state.Channel
+	if charm.CharmHub.Matches(curl.Schema) {
+		source = corecharm.CharmHub.String()
+		channel = &state.Channel{
+			Risk: "stable",
+		}
+	} else if charm.Local.Matches(curl.Schema) {
+		source = corecharm.Local.String()
+	}
+
+	b := base.MustParseBaseFromString("ubuntu@22.04")
+
+	platform := &state.Platform{
+		Architecture: corearch.DefaultArchitecture,
+		OS:           b.OS,
+		Channel:      b.Channel.String(),
+	}
+
+	return &state.CharmOrigin{
+		Source:   source,
+		Type:     "charm",
+		Revision: intPtr(curl.Revision),
+		Channel:  channel,
+		Platform: platform,
+	}
+}
+
+func intPtr(i int) *int {
+	return &i
+}
+
+func (s *statusUnitTestSuite) TestSubordinateUpgradingFrom(c *gc.C) {
+	factory, release := s.NewFactory(c, s.ControllerModelUUID())
+	release()
+
+	principalCharm := factory.MakeCharm(c, &testfactory.CharmParams{Name: "mysql", URL: "ch:amd64/mysql"})
```

Comment:
@ycliuhw Can you do a patch like https://github.com/juju/juju/pull/18441 to remove and document coverage for all the tests here that use `factory.MakeCharm`?